### PR TITLE
Add new ${github_notifications} variable to count all unread notifications

### LINKED ISSUE #636: Does a new ${github_notifications} variable makes sence ?
I'm willing to write a new variable `${github_notifications}` that parses all unread github notifications and counts them up. 

Does it makes sense to add it in conky :grey_question: 

edit: How often it should call the github API to obtain the current notifications status, I've set it to 60 seconds.

### PR DESCRIPTION
**Pull request by su8**
Out to resolve #636 

The user should generate a new and unique token if they want to use this variable - https://github.com/settings/tokens/new?scopes=notifications&description=conky-query-github

### CODE REVIEW

**Comment by lasers** [comment]
When I run this in `console`, it does not fetch every second, right? Just wanted to check. Knee-jerk reaction... And please tag me or something... so I can see `1`.  Also, I think maybe it should be `{github_notifications}` in case you or somebody else want to implement more `{github }` variables.

**Comment by su8** [comment]
@lasers ping, [dude where's my car](https://en.wikipedia.org/wiki/Dude,_Where%27s_My_Car%3F)

**Comment by su8** [comment]
It fetches the first time and caches the result, each sequential `print_github()` uses the cached result for next 60 times, once the variable `last_update` reaches **1** it queries github again and caches the result for next 60 `print_github()` calls.

**Comment by lasers** [comment]
```
Jesse: Dude! You got a tattoo!

Chester: So do you, dude! Dude, what does my tattoo say?

Jesse: "Sweet!" What about mine?

Chester: "Dude!" What does mine say?

Jesse: "Sweet!" What about mine?

Chester: "Dude!" What does mine say?

Jesse: "Sweet!" What about mine?

Chester: "Dude!" What does mine say?

Jesse: "Sweet!" What about mine?

Chester: "Dude!" But what does mine say?

Jesse: "Sweet!" What about mine?

Chester: "Dude!" What does mine say?

Jesse: "S - wee - t!" What about mine?

[later]

Chester: [angry] "Dude!" What does mine say?

Jesse: [screaming] "Sweet!"
```
Again! Ping me again. I checked my notification right away without checking conky first. :confused:

**Comment by su8** [comment]
```bash
conky.config = {
    out_to_console=true,
    out_to_x=false,
    github_token='8df3a97e3a8a93079b325e356c3cbcf1d556a894'
}
conky.text = [[
${github}
]]
```

**Comment by su8** [comment]
>If I don't specify token, does it print 0

It prints error msg:

```c
  if (0 == strcmp(github_token.get(*state).c_str(), "")) {
    NORM_ERR("${github} requires token. "
    "Go ahead and generate one "
    "https://github.com/settings/tokens/new?scopes=notifications&description=conky-query-github\n"
    "Insert it in conky.config = { github_token='TOKEN_SHA' }\n");
    return;
  }
````

**Comment by lasers** [comment]
@su8 What about invalid `github_token`? It should throw an error instead of printing `0`.

**Comment by su8** [comment]
>let github assign it with real one

That's hard to do it from C, we should point the url that generates tokes instead.

**Comment by lasers** [comment]
>we should point the url that generates tokes instead.

Do you mean print the URL or open an URL every 60 second? I think putting urls in conky windows / dialogs / bars / ncurses etc is bad idea. Just small error message would do (and maybe stop fetching the URL too?) Idk. 

Seems like we should hold off on this until you rewrite conky in python. Conky 3. 

Do what you think is sensible. Thoughts on `${github_notifications}`? We have 100 variables all in same place so I'm explicit with all variable names.

**Comment by su8** [comment]
`${github_notifications}` seems more accurate. As for the error msgs, we rely on the user to input correct notifications access token, I can add code that catches the two scenarios mentioned above and print the errors once every 60 seconds.

**Comment by lasers** [comment]
In case you didn't notice, we want `github.py`, not `github.cc`. Wink, wink.

**Comment by su8** [comment]
@lasers Why don't have a **dotfiles** repo ? I'd like to see your vim files.

**Comment by lasers** [comment]
@su8 Is this ready or are you working on a new commit?

**Comment by su8** [comment]
@lasers It's ready sir.

**Comment by lasers** [comment]
@su8 We don't want to catch invalid tokens?

**Comment by su8** [comment]
To remove the **Bad credentials** and **Missing the 'notifications' scope** checks ?

**Comment by lasers** [comment]
To stop printing misleading `0` when we entered an invalid token. I don't see this checks in?

File: src/common.cc [hunk 1]
```diff
@@ -673,6 +674,105 @@ void print_include(struct text_object *obj, char *p, unsigned int p_max_size) {
...
+              'd' == *(ptr+12)) { /* "Bad credentials" */
+            NORM_ERR("Bad credentials: generate a new token:\n" NEW_TOKEN);
+            snprintf(p, 80, "%s", "${github_notifications} Bad credentials. Generate a new token.");
+            skip = 1U;
+            break;
+          }
+          if ('M' == *(ptr+10) && 'i' == *(ptr+11) &&
+              's' == *(ptr+12)) { /* Missing the 'notifications' scope. */
+            NORM_ERR("Missing 'notifications' scope. Generate a new token\n" NEW_TOKEN);
+            snprintf(p, 80, "%s", "${github_notifications} Missing notifications scope. Generate a new token.");
```
Comment (su8) [inline src/common.cc]: Here's what it prints

File: src/common.cc [hunk 2]
```diff
@@ -673,6 +674,105 @@ void print_include(struct text_object *obj, char *p, unsigned int p_max_size) {
...
+        ++x;
+      }
+
+      if ('m' == *ptr && 'e' == *(ptr+1) &&
+          's' == *(ptr+2) && 's' == *(ptr+3)) { /* "message": */
+        if (z+13 < sz) { /* Verifying up to *(ptr+13) */
+          if ('B' == *(ptr+10) && 'a' == *(ptr+11) &&
+              'd' == *(ptr+12)) { /* "Bad credentials" */
+            NORM_ERR("Bad credentials: generate a new token:\n" NEW_TOKEN);
+            snprintf(p, 80, "%s", "${github_notifications} Bad credentials. Generate a new token.");
```
Comment (su8) [inline src/common.cc]: Here's what it prints

**Comment by su8** [comment]
But they will be left with the impression that something other failed, it's better to print which variable failed, so they can update it.

**Comment by lasers** [comment]
Prefix with `Github: ` instead?

**Comment by lasers** [comment]
`GitHub:`  <-- -Important.

**Comment by su8** [comment]
Thanks for adding it to the docs.

**Comment by su8** [comment]
I just rebased my commit to squash it and it seems I've deleted your docs commit :sob:

**Comment by lasers** [comment]
I think you can do `git pull` next time to pull in my commit.

**Comment by lasers** [comment]
Plz rebase. `git pull --rebase` + push.

**Comment by su8** [comment]
It says `Everything up-to-date`.

**Comment by su8** [comment]
Shall I use `git pull --rebase https://github.com/brndnmtthws/conky.git master` :question:

**Comment by lasers** [comment]
Ah yeah, that explains things.
```
git remote add upstream https://github.com/brndnmtthws/conky.git`
```
```
git for-each-ref --shell --format='git pull %(refname:lstrip=2) --set-upstream-to=upstream/master' refs/heads
```
See that command above? It's shell. Add `| sh` to run it live.

Then try things again.

**Comment by lasers** [comment]
@brndnmtthws Help! The tests have fallen, and the tests can't get up!

life_alert

**Comment by su8** [comment]
@lasers I'm leaving the squashing up to you, don't want to mess it up tjhis time.

**Comment by lasers** [comment]
@brndnmtthws Merge as-is.. without code smellz, plz.

**Pull request merged**